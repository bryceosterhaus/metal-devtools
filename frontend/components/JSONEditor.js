import Component, {Config} from 'metal-jsx';
import JSONEditor from 'metal-json-editor';

import {keys, isEqual} from 'lodash';

class MetalJSONEditor extends Component {
	arrowRenderer(expanded) {
		return <span class={expanded ? 'arrow down' : 'arrow right'} />;
	}

	handleOnChange(newValue) {
		const {onChange, type, value} = this.props;

		let changedData = {};

		keys(value).forEach(
			key => {
				const keyVal = value[key];
				const newKeyValue = newValue[key];

				if (!isEqual(keyVal, newKeyValue)) {
					changedData[key] = newKeyValue;
				}
			}
		);

		onChange(changedData, type);
	}

	render() {
		return (
			<JSONEditor
				arrowRenderer={this.arrowRenderer.bind(this)}
				data={this.props.value}
				onChange={this.handleOnChange.bind(this)}
			/>
		);
	}
}

MetalJSONEditor.PROPS = {
	config: Config.object(),
	onChange: Config.func(),
	type: Config.string(),
	value: Config.object()
};

export default MetalJSONEditor;
