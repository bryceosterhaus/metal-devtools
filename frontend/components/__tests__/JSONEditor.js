/* global IncrementalDOM */

jest.unmock('../JSONEditor');

import JSONEditor from '../JSONEditor';

describe('JSONEditor', () => {
	test('should render', () => {
		const component = new JSONEditor(
			{
				value: {}
			}
		);

		expect(component).toMatchSnapshot();
	});

	test('should return jsx element', () => {
		const component = new JSONEditor(
			{value: {}}
		);

		const initSpyVal = IncrementalDOM;

		IncrementalDOM = {
			elementVoid: jest.fn()
		};

		expect(component.arrowRenderer(true)).toMatchSnapshot();
		expect(component.arrowRenderer(false)).toMatchSnapshot();

		IncrementalDOM = initSpyVal;
	});

	test('should return changed data', () => {
		const spy = jest.fn();

		const component = new JSONEditor(
			{
				onChange: spy,
				value: {
					baz: {
						qux: 'test'
					},
					foo: 'bar'
				}
			}
		);

		const newVal = {
			baz: {
				qux: 'bar'
			},
			foo: 'bar'
		};

		expect(component.handleOnChange(newVal)).toMatchSnapshot();
		expect(spy).toHaveBeenCalled();
	});
});
