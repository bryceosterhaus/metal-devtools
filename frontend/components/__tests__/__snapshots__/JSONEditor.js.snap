exports[`JSONEditor should render 1`] = `<ul class="tree-container" style="list-style: none; margin: 0px;"></ul>`;

exports[`JSONEditor should return changed data 1`] = `undefined`;

exports[`JSONEditor should return jsx element 1`] = `undefined`;

exports[`JSONEditor should return jsx element 2`] = `undefined`;
